lib/ex_double_entry/schemas/account_balance.ex

```diff
@@ -105,8 +114,30 @@ defmodule ExDoubleEntry.AccountBalance do
 
   def lock_multi!(accounts, fun) do
     ExDoubleEntry.repo().transaction(fn ->
-      accounts |> Enum.sort() |> Enum.map(fn account -> lock!(account) end)
-      fun.()
+      locked_accounts =
+        accounts
+        |> Enum.sort()
+        |> Enum.reduce(from(record in AccountBalance), fn
+          %Account{scope: nil} = account, query ->
+            Ecto.Query.or_where(query, [
+              {:currency, ^account.currency},
+              {:identifier, ^account.identifier}
+            ])
+
+          %Account{} = account, query ->
+            Ecto.Query.or_where(query, [
+              {:currency, ^account.currency},
+              {:identifier, ^account.identifier},
+              {:scope, ^account.scope}
+            ])
+        end)
+        |> lock_cond(true)
+        |> ExDoubleEntry.repo().all()
+
+      accounts
+      |> Stream.map(fn account -> Enum.find(locked_accounts, &(&1.id == account.id)) end)
```

Comment:
Why do we need `Stream.map` here? Wouldn't it make things slower given the number of accounts?

Reply:
Actually, we don't need a stream here. It is just to cover the more general use-case of locking a lot of actions. In such a case, it should perform better in terms of memory.